nemo/utils/arguments.py

```diff
@@ -0,0 +1,105 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from argparse import ArgumentParser
+from typing import Any, Dict, List, Optional, Union
+
+
+def add_optimizer_args(
+    parent_parser: ArgumentParser,
+    optimizer: str = 'adam',
+    default_lr: float = None,
+    default_opt_args: Optional[Union[Dict[str, Any], List[str]]] = None,
+) -> ArgumentParser:
+    """Extends existing argparse with support for optimizers.
+
+    # Example of adding optimizer args to command line :
+    python train_script.py ... --optimizer "novograd" --lr 0.01 \
+        --opt_args betas=0.95,0.5 weight_decay=0.001
+
+    Args:
+        parent_parser (ArgumentParser): Custom CLI parser that will be extended.
+        optimizer (str): Default optimizer required.
+        default_lr (float): Default learning rate that should be overriden during training.
+        default_opt_args (list(str)): List of overriding arguments for the instantiated optimizer.
+
+    Returns:
+        ArgumentParser: Parser extended by Optimizers arguments.
+    """
+    if default_opt_args is None:
+        default_opt_args = []
+
+    parser = ArgumentParser(parents=[parent_parser], add_help=True, conflict_handler='resolve')
+
+    parser.add_argument('--optimizer', type=str, default=optimizer, help='Name of the optimizer. Defaults to Adam.')
+    parser.add_argument('--lr', type=float, default=default_lr, help='Learning rate of the optimizer.')
+    parser.add_argument(
+        '--opt_args',
+        default=default_opt_args,
+        nargs='+',
+        type=str,
+        help='Overriding arguments for the optimizer. \n Must follow the pattern : \n name=value separated by spaces.'
+        'Example: --opt_args weight_decay=0.001 eps=1e-8 betas=0.9,0.999',
+    )
+
+    return parser
+
+
+def add_scheduler_args(parent_parser: ArgumentParser) -> ArgumentParser:
+    """Extends existing argparse with default LR scheduler args.
+
+    Args:
+        parent_parser (ArgumentParser): Custom CLI parser that will be extended.
+
+    Returns:
+        ArgumentParser: Parser extended by LR Scheduler arguments.
+    """
```

Comment:
How come add_scheduler_args doesn't have the nice `optimize` `default_lr`, and `default_opt_args` that add_optimizer_args has to add default values?

Reply:
Those defaults are part of the optimizer not the scheduler, so it's included in the add optimizer args. But you are right, default scheduler args should take in a static dict / config object and set the defaults in argparse so that we have production ready schedulers. 

I'll submit another pr, this one was merged overnight.